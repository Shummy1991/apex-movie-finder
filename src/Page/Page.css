.movies-progress {
    margin-top: 144px;
    display: flex;
    justify-content: center;
}

.no-results {
    margin-top: 144px;
    text-align: center;
    color: white;
    font-size: 21px;
}