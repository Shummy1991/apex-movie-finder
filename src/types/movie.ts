export interface IGenre {
    name: string;
}

export interface IMovie {
    id: string;
    name: string;
    genres: IGenre[];
    score: number;
    overview: string;
    socialMedia: {
        imdb: string | null;
    };
    img: {
        url: string;
    } | null;
}
