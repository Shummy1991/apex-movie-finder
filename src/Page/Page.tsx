import { CircularProgress, Container } from "@mui/material";
import { useState } from "react";

import { searchMovies } from "../services";
import { searchSimilarMovies } from "../services/searchSimilarMovies";
import { IMovie } from "../types";

import Header from "./Header";
import Movies from "./Movies";
import "./Page.css";

const Page = () => {
  const [movies, setMovies] = useState<IMovie[] | null>(null);
  const [isLoading, setIsLoading] = useState(false);
  const handleSearch = async (query: string) => {
    setMovies(null);
    setIsLoading(true);
    try {
      const { data } = await searchMovies(query);
      setMovies(data.data.searchMovies);
    } catch (error) {
      console.log(error); // TODO: handle error
    } finally {
      setIsLoading(false);
    }
  };

  const handleGetRelatedMovies = async (id: string) => {
    setMovies(null);
    setIsLoading(true);
    const { data } = await searchSimilarMovies(id);
    setIsLoading(false);
    setMovies(data.data.movie.similar);
  }

  return (
    <Container maxWidth="sm">
      <Header handleSearch={handleSearch} isLoading={isLoading}/>
      {isLoading && (
        <div className="movies-progress">
          <CircularProgress />
        </div>
      )}
      {movies && movies.length > 0 && (
        <Movies movies={movies} handleGetRelatedMovies={handleGetRelatedMovies} />
      )}
      {movies && movies.length === 0 && (
        <div className="no-results">No results!</div>
      )}
    </Container>
  );
}

export default Page;
