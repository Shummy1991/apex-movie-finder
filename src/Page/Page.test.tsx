import "jest";
import { createEvent, fireEvent, render, waitFor } from "@testing-library/react";
import { act } from "react-dom/test-utils";

import { mockMovie } from "../test/movie";

import Page from "./Page";

jest.mock("../services/searchMovies", () => ({
    searchMovies: () => ({
        data: {
            data: {
                searchMovies: [mockMovie],
            },
        },
    }),
}));

jest.mock("../services/searchSimilarMovies", () => ({
    searchSimilarMovies: () => ({
        data: {
            data: {
                movie: {
                    similar: [mockMovie],
                },
            },
        },
    }),
}));

afterAll(() => {
    jest.clearAllMocks();
});

describe("MovieDetails", () => {
    it("matches snapshot", () => {
        const { asFragment } = render(<Page />)
        const component = asFragment();
        expect(component).toMatchSnapshot();
    });
    it("should handle search, fetching related movies", async () => {
        const { findByText, findAllByLabelText } = render(<Page />);
        const [textField] = await findAllByLabelText("Enter a movie title");
        const value = "the batman";
        fireEvent.change(
            textField,
            createEvent("input", textField, {
                target: {
                    value,
                }
            })
        );
        const searchButton = await findByText("Search");
        
        act(() => {
            fireEvent.click(searchButton);
        });
        const title = await findByText(mockMovie.name);
        await act(async () => {
            fireEvent.click(title);
            await waitFor(() => expect(findByText("Related movies")).toBeTruthy());
        });
        const related = await findByText("Related movies");
        fireEvent.click(related);
    });
});
