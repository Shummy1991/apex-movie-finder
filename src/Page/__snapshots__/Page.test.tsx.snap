// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetails matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
  >
    <header>
      <h3>
        Movie finder
      </h3>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-1s76rlx-MuiGrid-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for=":r0:"
              id=":r0:-label"
            >
              Enter a movie title
            </label>
            <div
              class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-gjbq6i-MuiInputBase-root-MuiFilledInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiFilledInput-input MuiInputBase-input css-10botns-MuiInputBase-input-MuiFilledInput-input"
                id=":r0:"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-qh4ozh-MuiGrid-root"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root Mui-disabled  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Search
          </button>
        </div>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
